textdb/textdb-storage/src/main/java/edu/uci/ics/textdb/storage/relation/RelationManager.java

```diff
@@ -0,0 +1,363 @@
+        String tableDirectory = dataStore.getDataDirectory();
+        Schema tableSchema = dataStore.getSchema();
+        IDField idField;
+                
+        if (! tableSchema.containsField(SchemaConstants._ID)) {
+            tableSchema = Utils.getSchemaWithID(tableSchema);
+        }
+        
+        ITuple insertionTuple = tuple;
+        // add "_id" to schema, and add ID field to tuple
+        if (! insertionTuple.getSchema().containsField(SchemaConstants._ID)) {
+            idField = new IDField(UUID.randomUUID().toString());
+            insertionTuple = getTupleWithID(tuple, idField);
+        } else {
+            idField = (IDField) tuple.getField(SchemaConstants._ID);
+        }
+
+        if (! tableSchema.equals(insertionTuple.getSchema())) {
+            throw new StorageException("Tuple's schema is inconsistent with table schema.");
+        }
+        
+        DataWriter dataWriter = new DataWriter(new DataStore(tableDirectory, tableSchema), luceneAnalyzer);
+        dataWriter.insertTuple(insertionTuple);
+
+        return idField;    
+    }
+
+    // delete a tuple by its id
+    public void deleteTuple(String tableName, IField idValue) throws StorageException {
+        try {
+            DataStore tableDataStore = getTableDataStore(tableName);
+            String tableAnalyzerString = getTableAnalyzer(tableName);
+            Analyzer luceneAnalyzer = LuceneAnalyzerConstants.getLuceneAnalyzer(tableAnalyzerString);
+            
+            Query tupleIDQuery = new TermQuery(new Term(SchemaConstants._ID, idValue.getValue().toString()));
+            DataWriter tableDataWriter = new DataWriter(tableDataStore, luceneAnalyzer);
+            tableDataWriter.deleteTuple(tupleIDQuery);   
+        } catch (DataFlowException e) {
+            throw new StorageException(e);
+        }   
+    }
+
+    // update a tuple by its id
+    public void updateTuple(String tableName, ITuple newTuple, IDField idValue) throws StorageException {
+        if (getTuple(tableName, idValue) == null) {
+            throw new StorageException(
+                    String.format("Tuple with id %s doesn't exist in table %s.", idValue, tableName));
+        }
+
+        ITuple newTupleWithID = getTupleWithID(newTuple, (IDField) idValue);
+        if (newTupleWithID.getField(SchemaConstants._ID) != (IDField) idValue) {
+            throw new StorageException("New tuple's ID is inconsistent with idValue.");
+        }
+        
+        deleteTuple(tableName, idValue);
+        insertTuple(tableName, newTupleWithID);
+    }
+
+    // get a tuple by its id
+    public ITuple getTuple(String tableName, IDField idValue) throws StorageException {
+        DataStore tableDataStore = getTableDataStore(tableName);
+        
+        Query tupleIDQuery = new TermQuery(new Term(SchemaConstants._ID, idValue.getValue().toString()));
```

Comment:
Add a TODO to indicate the fact that if the caller calls this function many times, many java objects (TermQuery and Reader) will be created.  An improvement is to create them once and reuse them for each call so that we can avoid creating too many java objects.